kernel/src/checkpoint/mod.rs

```diff
@@ -0,0 +1,485 @@
+//!
+//! ```text
+//! +------------------+-------------------------------+
+//! | Table Feature    | Resulting Checkpoint Type     |
+//! +==================+===============================+
+//! | No v2Checkpoints | Single-file Classic-named V1  |
+//! +------------------+-------------------------------+
+//! | v2Checkpoints    | Single-file Classic-named V2  |
+//! +------------------+-------------------------------+
+//! ```
+//!
+//! Notes:
+//! - Single-file UUID-named V2 checkpoints (using `n.checkpoint.u.{json/parquet}` naming) are to be
+//!   implemented in the future. The current implementation only supports classic-named V2 checkpoints.
+//! - Multi-file V2 checkpoints are not supported yet. The API is designed to be extensible for future
+//!   multi-file support, but the current implementation only supports single-file checkpoints.
+//! - Multi-file V1 checkpoints are DEPRECATED.
+//!
+//! ## Example: Writing a classic-named V1/V2 checkpoint (depending on `v2Checkpoints` feature support)
+//!
+//! TODO(seb): unignore example
+//! ```ignore
+//! let path = "./tests/data/app-txn-no-checkpoint";
+//! let engine = Arc::new(SyncEngine::new());
+//! let table = Table::try_from_uri(path)?;
+//!
+//! // Create a checkpoint writer for the table at a specific version
+//! let mut writer = table.checkpoint(&engine, Some(2))?;
+//!
+//! // Retrieve checkpoint data
+//! let checkpoint_data = writer.get_checkpoint_info()?;
+//!
+//! /* Write checkpoint data to file and collect metadata about the write */
+//! /* The implementation of the write is storage-specific and not shown */
+//! /* IMPORTANT: All data must be written before finalizing the checkpoint */
+//!
+//! // Finalize the checkpoint by writing the _last_checkpoint file
+//! writer.finalize_checkpoint(&engine, &checkpoint_metadata)?;
+//! ```
+//!
+//! This module, along with its submodule `checkpoint/log_replay.rs`, provides the full
+//! API and implementation for generating checkpoints. See `checkpoint/log_replay.rs` for details
+//! on how log replay is used to filter and deduplicate actions for checkpoint creation.
+use crate::{
+    actions::{
+        schemas::GetStructField, Add, Metadata, Protocol, Remove, SetTransaction, Sidecar,
+        ADD_NAME, METADATA_NAME, PROTOCOL_NAME, REMOVE_NAME, SET_TRANSACTION_NAME, SIDECAR_NAME,
+    },
+    expressions::Scalar,
+    path::ParsedLogPath,
+    schema::{DataType, SchemaRef, StructField, StructType},
+    snapshot::Snapshot,
+    DeltaResult, Engine, EngineData, Error, EvaluationHandlerExtension,
+};
+use log_replay::{checkpoint_actions_iter, CheckpointData};
+use std::{
+    cell::RefCell,
+    rc::Rc,
+    sync::{Arc, LazyLock},
+    time::{Duration, SystemTime, UNIX_EPOCH},
+};
+use url::Url;
+
+mod log_replay;
+#[cfg(test)]
+mod tests;
+
+/// Schema for extracting relevant actions from log files during checkpoint creation
+static CHECKPOINT_READ_SCHEMA: LazyLock<SchemaRef> = LazyLock::new(|| {
```

Comment:
nit: maybe rename to something so we don't confuse this as the schema we are supposed to _read_ checkpoints with
```suggestion
static CHECKPOINT_ACTIONS_SCHEMA: LazyLock<SchemaRef> = LazyLock::new(|| {
```